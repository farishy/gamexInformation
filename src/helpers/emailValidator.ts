/**
 * validateEmail
 * ---
 *
 * ---
 *
 * Returns a boolean to indicate an email validity.
 *
 * @param email
 * The email string to be validated.
 */

export default function validateEmail(email: string) {
  const emailRegex =
    /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  return emailRegex.test(email);
}
