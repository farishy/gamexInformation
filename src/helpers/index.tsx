import formatMdn from './mdnFormatter';
import validateMdn from './mdnValidator';
import sanitizeMdn from './mdnSanitizer';
import validateEmail from './emailValidator';
import validatePassword from './passwordValidator';
import validateGeneralMdn from './mdnGeneralValidator';
import validateName from './nameValidator';

export {
    validateMdn,
    formatMdn,
    sanitizeMdn,
    validateEmail,
    validatePassword,
    validateGeneralMdn,
    validateName,
  };