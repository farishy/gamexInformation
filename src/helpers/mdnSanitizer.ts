/**
 * mdnFormatter
 * ---
 *
 * ---
 *
 * Convert phone numbers from '+62-856-2101-7923' to '6285621017923'.
 *
 * @param mdn
 * The phone number in string.
 */
export default function sanitizeMdn(mdn: string) {
  if (!mdn) {
    return mdn;
  }

  const _mdnNumberOnly = mdn.replace(/[+-]/g, '');
  const _allowedPrefixes = ['628', '08', '8'];
  let _sanitizedMdn = '';

  _allowedPrefixes.forEach((prefix) => {
    if (_mdnNumberOnly.indexOf(prefix) === 0) {
      _sanitizedMdn = _mdnNumberOnly.replace(prefix, '6288');
    }
  });

  return _sanitizedMdn;
}
