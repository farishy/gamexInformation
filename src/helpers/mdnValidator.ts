/**
 * validatePhoneNumber
 * ---
 *
 * ---
 *
 * Returns a boolean to indicate a phone number's validity.
 *
 * @param mdn
 * The phone number string to be validated.
 */

let validateMdn = (mdn: string) => {
  return /^(\+628|628|08|8)\d{5,30}/.test(mdn);
};

export default validateMdn;
