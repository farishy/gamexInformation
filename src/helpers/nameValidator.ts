/**
 * validateFullname
 * ---
 *
 * ---
 *
 * Returns a boolean to indicate a full name's validity.
 *
 * @param fullname
 * The full name string to be validated.
 */

 let validateName = (name: string) => {
    return /^[a-zA-Z]+(([',. -][a-zA-Z ])?[a-zA-Z]*)*$/g.test(name);
  };
  
  export default validateName;
  