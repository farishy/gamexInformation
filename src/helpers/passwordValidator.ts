/**
 * validatePassword
 * ---
 *
 * ---
 *
 * Returns a boolean to indicate a password's validity.
 *
 * @param password
 * The password string to be validated.
 */

let validatePassword = (password: string) =>
  /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).{8,}$/.test(password);

export default validatePassword;
