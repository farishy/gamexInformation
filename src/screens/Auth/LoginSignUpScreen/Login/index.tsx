import { anyTypeAnnotation } from '@babel/types';
import AsyncStorage from '@react-native-async-storage/async-storage';
import React, { useEffect, useState } from 'react';
import { StyleSheet, Text, View, ScrollView, Keyboard, Alert } from 'react-native';
import CustomButton from '../../../../components/CustomButton';
import CustomButtonCircle from '../../../../components/CustomButtonCircle';
import CustomTextInput from '../../../../components/CustomTextInput';
import HyperLink from '../../../../components/HyperLink';
import {validateEmail, validatePassword} from '../../../../helpers';

const Login = (props : any) =>{

    const googleLogo = '../../../../assets/images/google_logo.png'
    const facebookLogo = '../../../../assets/images/facebook_logo.png'
    const [email, setEmail] = useState<string>('');
    const [password, setPassword] = useState<string>('');
    const [emailValid, setEmailValid] = useState<boolean>(false);
    const [passwordValid, setPasswordValid] = useState<boolean>(false);
    const [buttonStats, setbButtonStats] = useState<boolean>(false);
    const [emailErrMessage, setEmailErrMessage] = useState<string>('');
    const [passwordErrMessage, setPasswordErrMessage] = useState<string>('');

    const emailValidation = (email : any) => {
        if(email.length < 1){
            setEmailValid(false);
            setEmailErrMessage('');
        }else if(validateEmail(email)){
            setEmailValid(true);
            setEmailErrMessage('');
        }else{
            setEmailValid(false);
            setEmailErrMessage('Email tidak valid');
        }
    }

    const passwordValidation = (password : any) => {
        if(password.length < 1){
            setPasswordValid(false);
            setPasswordErrMessage('');
        }else if(validatePassword(password)){
            setPasswordValid(true);
            setPasswordErrMessage('');
        }else{
            setPasswordValid(false);
            setPasswordErrMessage('Password terdiri 8 karakter dengan numerik, huruf besar dan huruf kecil');
        }
    }

    const doLogin = async () => {
        let user = await AsyncStorage.getItem('user');
        let _data = user !== null ? JSON.parse(user) : {};
        if(email===_data.email && password===_data.password){
            Alert.alert(
                "Status Login",
                "Berhasil",
                [
                    { text: "OK", onPress: () => {
                        setEmail('');
                        setPassword('');
                    }}
                ]
            )
        }else{
            Alert.alert(
                "Status Login",
                "Gagal",
                [
                    { text: "OK", onPress: () => {
                        setEmail(email);
                        setPassword(password);
                    }}
                ]
            )
        }
    }

    useEffect(() => {
        emailValidation(email);
        passwordValidation(password);

        const allValid = emailValid && passwordValid;
        if (allValid) {
            setbButtonStats(true);
        } else {
            setbButtonStats(false);
        }
    }, [email, password])

    return(
        <View style={styles.container}>
            <ScrollView>
                <View style={{marginHorizontal:20, marginTop:10}}>
                    <CustomTextInput
                        placeholder='Email'
                        icon='emoticon-happy'
                        label='Email'
                        value={email}
                        onChangeText={(text: React.SetStateAction<string>) => setEmail(text)}
                        errorMessage={emailErrMessage}
                        testID='login-input-email'
                    />
                    <CustomTextInput
                        label='Kata Sandi'
                        placeholder='Kata Sandi'
                        icon='lock'
                        secureTextEntry={true}
                        value={password}
                        onChangeText={(text: React.SetStateAction<string>) => setPassword(text)}
                        errorMessage={passwordErrMessage}
                        testID={'login-input-password'}
                    />
                    <HyperLink 
                        label='Lupa Kata Sandi ?'
                        style={{marginBottom:30, height:20,  alignSelf:'flex-end', fontWeight:'bold'}}
                        color='#398FCE'
                        fontWeight='bold'
                        testID='btn-lupaPassword'
                    />
                    <CustomButton style={{marginBottom:20,}} label='MASUK' disabled={!buttonStats} onPress={doLogin} testID='btn-masuk'/>
                    <Text style={{textAlign:'center', fontSize:16, paddingBottom:15, color:'black'}}>Atau masuk menggunakan</Text>
                    <View style={{flexDirection:'row', justifyContent:'center', marginBottom:20}}>  
                        <CustomButtonCircle source={require(googleLogo)} style={styles.CustomButtonCircle} testID='btn-google'/>
                        <CustomButtonCircle source={require(facebookLogo)} size={30} style={styles.CustomButtonCircle} testID='btn-facebook'/>
                    </View>
                </View>
            
            </ScrollView>
            
        </View>
    )
};

const styles = StyleSheet.create({
    container:{
        flex:1,
    },
    CustomButtonCircle:{
        padding:10,
    }

});

export default Login