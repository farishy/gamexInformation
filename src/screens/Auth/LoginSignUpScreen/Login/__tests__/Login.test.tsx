import React from 'react'
import Login from '../../Login';
import {render, fireEvent} from '@testing-library/react-native';
import {validateEmail, validatePassword} from '../../../../../helpers';
import { Alert } from 'react-native';

describe('Login Screen', ()=>{
  const { getByTestId, getByText } = render(<Login />);
  const emailInputText = getByTestId('login-input-email');
  const passwordInputText = getByTestId('login-input-password');
  const lupaPasswordHyperlink = getByTestId('btn-lupaPassword');
  const buttonMasuk = getByTestId('btn-masuk');
  const anotherLoginText = getByText('Atau masuk menggunakan');
  const buttonGoogle = getByTestId('btn-google');
  const buttonFacebook = getByTestId('btn-facebook');

  it('should rendering the login screen components', ()=>{
    expect({
      emailInputText, 
      passwordInputText, 
      lupaPasswordHyperlink, 
      buttonMasuk, 
      anotherLoginText, 
      buttonGoogle, 
      buttonFacebook
    }).toBeDefined;
  });
  it('should helpers is working too', ()=>{

    const valueEmail = 'muhammadfaris';
    const valuePassword = 'Testing123';

    const expectValueEmail = validateEmail(valueEmail);
    const expectValuePassword = validatePassword(valuePassword);

    expect(expectValueEmail).not.toBeTruthy();
    expect(expectValuePassword).toBeTruthy();

  });
  it('should change value of input', ()=>{
    const onEventMock = jest.fn();
    const { getByTestId } = render(
      <Login handleChangeText={onEventMock} />
    );
    fireEvent.changeText(getByTestId('login-input-email'), 'onChangeText', 'Muhammadfaris9d@gmail.com');
    fireEvent.changeText(getByTestId('login-input-password'), 'onChangeText', 'Testing1234');
    
    expect(getByTestId('login-input-email')).not.toBeNull();
    expect(getByTestId('login-input-password')).not.toBeNull();

  });
  test('button masuk have disabled', ()=>{
    const onEventMock = jest.fn();
    const { getByTestId } = render(
      <Login onPress={()=> onEventMock}/>
    );
    fireEvent.press(getByTestId('btn-masuk'));
    expect(onEventMock).not.toBeCalled();
  });
  test('login flow', async ()=>{
    const alertMock = jest.spyOn(Alert,'alert').mockImplementation();
    const { getByPlaceholderText, getByText } = render(
      <Login />
    );
    await fireEvent.changeText(getByPlaceholderText('Email'), 'muhammadfaris9d@gmail.com');
    await fireEvent.changeText(getByPlaceholderText('Kata Sandi'), 'Testing1234')
    await fireEvent.press(getByText('MASUK'));
  
    expect(alertMock).toHaveBeenCalled();

  })
});



// import 'react-native'
// import React from 'react'
// import Adapter from 'enzyme-adapter-react-16';
// import {shallow, configure, mount} from 'enzyme'
// import renderer from 'react-test-renderer'
// import Login from '../../Login'
// import CustomTextInput from '../../../../../components/CustomTextInput';
// configure({adapter: new Adapter(), disableLifecycleMethods: true})

// const loginWrapper = shallow(<Login />)

// describe('Login Screen', () => {
//     it('should renders correctly', () => {
//       renderer.create(<Login />)
//     })
   
//     it('should renders `Login Screen` module correctly', () => {
//       expect(loginWrapper).toMatchSnapshot()
//     })
   
//     describe('Check component input file', () => {
//       it('should create TextInput`', () => {
//         expect(loginWrapper.find('TextInput').exists())
//       })
   
//       it('should create button component', () => {
//         expect(loginWrapper.find('TouchableOpacity').exists())
//       })
   
//       describe('Mount component', () => {
//         describe('Event Test', () => {
//           it('should change text when `Input` Change', () => {
//             const mockCallBack = jest.fn()
//             const props = {
//               label: 'Test Label',
//               input: {
//                 mockCallBack
//               },
//               value: 'Hello world',
//           }
//           const wrapper = mount(<CustomTextInput {...props}/>);
//           const event = {
//             target: {
//                 value: 'This is just for test'
//             }
//         }
//             wrapper.find('[data-testID={"input-email"}]').simulate('change', event)
//             expect(mockCallBack).toHaveBeenCalledWith('This is just for test');
//             // loginWrapper.find('[data-testID={"input-email"}]').simulate('change')
//             // expect(mockCallBack.mock.calls.length)
//           })
//           it('should click when lupa password Pressed', () => {
//             const mockCallBack = jest.fn()
//             loginWrapper.find('[data-testID="btn-lupaPassword"]').simulate('press')
//             expect(mockCallBack.mock.calls.length)
//           })
//         })
//       })

//     })
//   })