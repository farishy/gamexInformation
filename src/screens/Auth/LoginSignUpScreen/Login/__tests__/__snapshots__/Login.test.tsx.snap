// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Screen should renders \`Login Screen\` module correctly 1`] = `ShallowWrapper {}`;
